### PR TITLE
fix(poller): pool-aware claim clamp and congestion-safe fail path

### PR DESCRIPTION
**Pull request by bodymindarts**
Implements pool-aware claiming and fail-path safety for both dispatch paths (batch and single-job), per `job-dev:handoff-pool-aware-claiming-and-fail-path.md` (drua library): work items A (pool-aware claim clamp), B (congestion classification), and C (terminal-write safety).

## The problem

Under shared-pool pressure, the poller used to claim work the pool could not serve. The consequences compounded: claimed rows stranded `running` with no connection to run on (recoverable only by the lost-handler, one `job_lost_interval` later), `PoolTimedOut` failures burned `RetrySettings` attempts on jobs that were never broken, and the fail write itself — the last chance to record a disposition — could fail to acquire a connection from the very pool whose exhaustion caused the failure.

## Mechanism

**A — pool-aware claim clamp** (`poller.rs`, `registry.rs`, `config.rs`). `pool_connection_headroom` reads live shared-pool headroom (`max_connections - in_use`). `JobPoller::pool_unit_budget` turns that into a **dispatch-unit** budget — `floor(headroom / connections_per_job)` — fed into `JobRegistry::plan_claim`:

- A *unit* is a dispatch, not a row: a batched type costs `ceil(row_limit / max_batch_size)` units (the chunk count `dispatch_batches` actually produces), a plain type one unit per row. Traced through `batched.rs`/`current.rs`: the crate's own claim/dispatch machinery never holds more than one connection per dispatch — which is why **1.0 is the default price**, and only the default: `JobPollerConfig::connections_per_job` (new, `f64`, serde default `1.0`, validated finite in `(0, 100]`) lets a deployment tune it fractionally, because the operator knows their workload's connection shape where the crate cannot. Mostly-connection-free runners: `0.5` admits twice the headroom; fan-out-heavy runners: `1.5`/`2.0` admit less. Under-admission mistakes at any price are absorbed by B.
- The budget is spent **smallest-unit-demand-first**, never in `HashMap` iteration order — under a scarce budget, registration-order spending would let a large uncapped type exhaust the budget before a 1-slot type gets a turn (a real starvation bug caught by live-PG testing during development).
- **Zero budget claims nothing** — rows stay `pending`, claimable by peer instances with healthy pools. (An interim revision floored the budget at one unit; Bugbot correctly flagged that a row claimed by a saturated instance sits `running`, locked away from every healthy peer for up to the pool's ~30s acquire timeout — the floor's single-instance wake argument inverts multi-instance.) Instead, a clamped-to-zero poll arms a **pool-headroom waiter**: one background task re-checking the budget on a 10ms→1s doubling backoff and waking the poll loop the moment a whole unit's headroom is back — fine-grained enough to catch millisecond-scale connection churn a fixed retry interval would miss, one cheap in-process read per second once genuinely stuck. `MAX_WAIT` (60s) remains only as the lost-wake fallback.
- The **spawn/recycle short-circuit (`ClaimHook`) is gated by the same budget** — it previously claimed with no headroom check at all, so a same-instance spawn on a saturated instance stranded its row exactly the way the poll-path floor did (found by the new waiter test spawning through the fast path). Zero budget: recycled reservations release ordinarily (`UnitReservation::Drop`, which now wakes the poll loop like every other release path), fresh demand stays pending for the pool-aware poll. The claim query's SQL, per-type windowing, and the dedicated poll connections are untouched.

**B — congestion classification** (`finalizer.rs`, both dispatchers). Runner errors are classified by `Finalizer::maybe_reclassify` (a source-chain walk for `sqlx::Error::PoolTimedOut`, done *before* stringification — both dispatchers' error branches are one line: `Err(finalizer.maybe_reclassify(e))`) into `JobError::PoolCongestion` vs real errors. Congestion routes to `Disposition::Congestion` instead of the `RetryPolicy` fail path:

- Rows go back to `pending` at now + 2s ± 1s jitter, **`attempt_index` unchanged** — congestion carries no evidence the job is broken, so it must not walk the job toward `max_attempts`. Preserving the attempt is also load-bearing for the poller's retry-solo rule (`attempt > 1` dispatches alone, never batched): resetting it would silently change dispatch shape. See `congestion_reschedule_keeps_job_batchable`.
- A fresh `CongestionRescheduled` entity event (not `ExecutionErrored`) tracks a consecutive-congestion streak; past 10, a WARN fires so "stuck in congestion forever" is observable. Non-punitive stays non-punitive — it's a signal, not a cap.
- A congestion reschedule deliberately skips the completion-recycle short-circuit (`try_recycle_own_type`): that path claims and re-dispatches the same type immediately with no headroom check — exactly backwards right after evidence the pool is unhealthy. The freed unit releases through the ordinary pool-aware poll instead.

**C — terminal-write safety + the `Finalizer`** (`finalizer.rs`, `poller.rs`, both dispatchers). The dedicated poll pool is renamed `internal_pool` and sized to 4 connections, and **every** end-of-job disposition — complete, fail/retry, runner reschedule, congestion, rescue — flows through one five-phase write in `finalizer.rs` (`Finalizer::finalize_in_op`: load entities → decide + push events → `(queue_id, id)`-ordered row writes → hook registrations → entity updates), with `Disposition::{Complete, Fail, Fresh, Congestion}` as the four SET-list/event variants. A solo job is a batch of one; the terminal delete that used to exist in three copies is one parameterized query (`retains_state` covers keyed `inherits_state`; batched types are never keyed). Keyed and resident jobs inherit all of it via the ordinary dispatcher funnel.

**Pool policy**, implemented once in `Finalizer::finalize`: the *first* attempt of a dispatcher-owned disposition transaction runs on the **shared pool** when it has live headroom — keeping the small internal pool unloaded in the healthy case — under a hard **1s deadline** independent of the pool's ~30s acquire timeout; **any** first-attempt failure (including `PoolTimedOut`, which is not a retryable abort) switches to the **internal pool**, where transient aborts (`is_retryable_conflict`, `TX_ABORT_MAX_ATTEMPTS` in `error.rs`) retry. Zero shared headroom skips the doomed shared attempt outright. A starved shared pool can never prevent the crate from recording a job's disposition — and this also upgraded the batch *seal*, which previously ran shared-only with no internal fallback at all.

Both dispatchers now have the full **rescue** last resort: a disposition-write failure hands the claimed rows back `pending` (`Disposition::Fresh` at now, attempt reset) with a `claim_disposition` (`rescued`/`leaked`) recorded on the execution span — the lost-handler is a true backstop, not the primary recovery path.

## Consumer impact

- Worst-case simultaneous job-side connection demand is `N + 1` (`N` = sum of effective per-process concurrency across types, `+1` for the notification listener), plus `internal_pool`'s 4 — but the claim clamp means steady-state usage tracks live headroom rather than the theoretical max. Recomputed for lana-bank: worst case 30, inside its 40-connection pool. `connections_per_job` scales the admission side of this without touching the worst case.
- Multi-instance deployments get the peer-takeover property for free: an instance under pool pressure stops claiming entirely, so backlog drains through whichever instances are healthy instead of stranding behind the saturated one.
- If job traffic and application traffic genuinely compete over a shared pool, the intended fix is deployment-shaped: hand `Jobs` a dedicated pool (a one-line usage change). The claim clamp is self-protection against over-claiming, not an app-traffic priority mechanism — documented on `pool_unit_budget`.
- No migration: congestion tracking is a new JSONB-stored `JobEvent::CongestionRescheduled` variant, not a new column. No baseline migration touched.

## Test plan

- New live-PG coverage, batch and solo mirrored: `pool_congestion.rs` / `pool_congestion_solo.rs` (attempt preserved across a congestion hit; job stays batchable), `pool_terminal_write_safety.rs` / `pool_terminal_write_safety_solo.rs` (terminal write survives full shared-pool exhaustion), plus a unit test pinning `pool_connection_headroom`'s live tracking.
- `pool_headroom_waiter.rs`: `saturated_instance_leaves_rows_pending_for_healthy_peer` pins the multi-instance regression directly (two `Jobs` instances, A saturated and polling first/alone, B healthy — B must complete the job, A must never touch it), `waiter_wakes_poller_when_headroom_returns` pins recovery (one connection freed → the job runs promptly, not after `MAX_WAIT`). Both state-polled, no blind sleeps; the drain helper requires *stable* saturation (holds everything but the `LISTEN` connection and verifies the pool reports fully checked out) since a plain acquire-until-timeout drain races the poller monitors.
- `unit_budget` unit tests (1.0 identity, 0.5/1.5/2.0 direction, floor edge cases incl. `floor(1/1.5) = 0`), `connections_per_job` serde-default and builder-validation tests.
- Test pool stays at its pre-feature `max_connections(5)`; repeated full-suite runs clean (8/8 at the time of the sizing decision, and again after each follow-up commit). Claim-clamp causality was established by controlled A/B against an isolated `origin/main` clone rather than assumed — the branch flakes *less* than main's own baseline (main: 3/8 on the known `batch_error_retries_every_item_and_retries_run_alone` suite-load flake, owned by 01a0293e).
- `cargo clippy --all-targets --all-features -- -D warnings` clean; `nix flake check` clean; `.sqlx` regenerated (net: one entry removed — the solo congestion UPDATE, absorbed by the shared N-row query).

## Related, explicitly out of scope

- **Per-TYPE connection cost (named, not built).** `connections_per_job` prices the whole deployment with one tunable factor; a per-type or per-runner declared cost (defaulting to it) remains a reasonable follow-up for a mixed workload where one type's runners are much heavier than the rest. Undercharging any type stays cheap either way — a rare real `PoolTimedOut` costs a short non-punitive reschedule, not a burned attempt.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches claim admission, retry/disposition writes, and shared vs internal pool usage—the core job-execution path. Bugs here can strand running rows, skip retries, or deadlock under contention.
> 
> **Overview**
> Stops the poller from claiming jobs the shared pool cannot actually run, and makes fail/complete writes survive pool exhaustion instead of stranding rows as `running` until the lost-handler.
> 
> **Admission** now budgets dispatch *units* from live pool headroom (`floor(headroom / connections_per_job)`, default `1.0`, operator-tunable). Zero budget claims nothing so peers can take the work, and a backoff waiter wakes the poll loop when a full unit of headroom returns. The same clamp applies to completion-recycle / short-circuit claims.
> 
> **Congestion** (`PoolTimedOut`) is no longer a retry-policy failure. Jobs go back to `pending` after a short jittered delay with `attempt_index` unchanged (so they stay batchable), recorded as `CongestionRescheduled`. A consecutive-streak WARN surfaces jobs stuck in congestion.
> 
> **Disposition** for both solo and batch jobs is centralized in `Finalizer`: complete / fail / fresh / congestion share one ordered write path, promote/notify hooks, and abort retries. Dispatcher-owned writes try the shared pool first (1s acquire cap, skip if no headroom) then fall back to a 4-connection `internal_pool`. If a disposition write still fails, claimed rows are rescued as `Fresh` rather than leaking until lost-job sweep.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e034aca7119f1424ae7cb817a34e05220617913c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/poller.rs [hunk 1]
```diff
@@ -348,15 +443,17 @@ impl JobPoller {
             span.record("n_jobs_to_start", 0);
             return Ok(MAX_WAIT);
         };
-        let plan = self.registry.plan_claim(n_jobs_to_poll);
+        let unit_budget = self.pool_unit_budget();
+        let plan = self.registry.plan_claim(n_jobs_to_poll, unit_budget);
+        span.record("n_claim_clamped_by_pool", plan.clamped_by_pool);
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/poller.rs:594]: Confirmed and fixed in cc6f059: `plan.clamped_by_pool` (already computed, already recorded on the span) now selects a short `POOL_CLAMP_RETRY` (2s) instead of `MAX_WAIT` (60s) when a poll had due work but the pool clamp reduced every type's claim to zero units. Genuinely-idle polls (no due work at all) keep the 60s backstop.

File: src/batch_dispatcher.rs [hunk 2]
```diff
@@ -906,6 +996,142 @@ impl BatchDispatcher {
...
+                let repo = self.terminal_write_repo();
+                let mut op = repo.begin_op_with_clock(&self.clock).await?;
+                let now = op.maybe_now().unwrap_or_else(|| self.clock.now());
+                let jitter_ms = rng().random_range(-CONGESTION_JITTER_MS..=CONGESTION_JITTER_MS);
+                let scheduled_at =
+                    now + chrono::Duration::milliseconds(CONGESTION_DELAY_MS + jitter_ms);
+                let streak = self
+                    .congestion_reschedule_in_op(&mut op, message.clone(), scheduled_at)
+                    .await?;
+                self.try_recycle_own_type(&mut op);
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/batch_dispatcher.rs]: Confirmed and fixed in cc6f059: removed the `try_recycle_own_type` call from `reschedule_congestion`. That call was handing the freed unit straight to `ClaimHook`, which claims and dispatches more due work of the same type immediately with no reference to `pool_unit_budget` at all -- right after a congestion signal specifically, that could feed a tight claim-dispatch-fail-reschedule loop with no backoff. The unit now releases through the ordinary `Drop`/`batch_completed` path instead, picked back up by the next pool-aware poll (which checks headroom fresh). Confirmed `JobDispatcher::reschedule_congestion` (single-job path) never had the equivalent short-circuit call in the first place, so nothing to fix there.

File: src/poller.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/poller.rs:594]: Update in a6bbdd9: the 2s `POOL_CLAMP_RETRY` from cc6f059 is superseded by min-admission — `pool_unit_budget` is now floored at one unit (`max(headroom, 1)`), so a clamped poll claims one unit instead of sleeping on a timer. The blocked `acquire()` itself becomes the wake signal (runs the instant a connection frees, vs up to 2s of stat staleness), and a job that genuinely cannot acquire takes the non-punitive congestion reschedule. Pile-up under sustained saturation is bounded at `min_jobs_per_process` by the tracker gate. The timer, its constant, and the clamped-empty-poll branch are deleted; see the PR body ("Follow-up: min-admission floor of one unit") for the full reasoning.

File: src/poller.rs [hunk 3]
```diff
@@ -334,11 +388,96 @@ impl JobPoller {
...
+    ///   `next_batch_size` stops the poller at `min_jobs_per_process`
+    ///   blocked acquires, recycling from there on the pool's
+    ///   `acquire_timeout` cadence.
+    ///
+    /// The floor means jobs keep competing for a saturated pool rather than
+    /// yielding it. If that pool is shared with an application and the
+    /// competition hurts, the fix is deployment-shaped, not more poller
+    /// machinery: hand `Jobs` its own pool.
+    fn pool_unit_budget(&self) -> usize {
+        pool_connection_headroom(self.repo.pool()).max(1)
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/poller.rs]: This is a compelling argument to revert to the wait approach. 

I think we should do a third option. 

If we don’t have any headroom spawn a time task that wakes the poller with a notification once one is free. Similar to this test code:

async fn settle(pool: &PgPool, expected_idle: u32) {
        for _ in 0..1000 {
            if pool.num_idle() as u32 == expected_idle {
                return;
            }
            tokio::task::yield_now().await;
        }
    }
    
Just that the condition is idle >=1. 

And loop with a back off. 

It should start the loop with like 10ms and back off gradually to 1s or so. 

2s fix wait seems very long and will miss a lot of connection churn.

[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/poller.rs]: Implemented in 3aa9295 (+ 74d3983 for the config factor). Bugbot's multi-instance point is correct and the floor is gone: zero headroom now claims nothing, so rows stay `pending` where a healthy peer takes them — pinned by a new two-instance test (`saturated_instance_leaves_rows_pending_for_healthy_peer`: A saturated and polling first/alone, B healthy; B must complete the job, A must never touch it).

The waiter is as described: a clamped-to-zero poll arms one background task that re-checks the budget on a **10ms doubling-to-1s backoff** and wakes the poll loop via the tracker's notify the moment a unit's headroom is back. Doubling because pool exhaustion is usually query-duration-scale: the 10/20/40/80ms early checks catch millisecond churn the fixed 2s wait missed, while the geometric growth settles a genuinely stuck pool at one cheap in-process counter read per second (~8 checks to ceiling). Real backoff, not `yield_now`; at most one waiter (atomic arm flag, disarm-before-wake so the woken poll can re-arm); `Weak` on the poller so it dies with it. Condition is `pool_unit_budget() > 0` rather than `num_idle() >= 1` — headroom also counts never-opened capacity, and with `connections_per_job > 1.0` one idle connection can still be a zero budget (waking on it would spin).

Implementing this surfaced two adjacent holes, both fixed in the same commit: the spawn/recycle short-circuit (`ClaimHook::pre_commit`) claimed with **no headroom check at all** — a same-instance spawn on a saturated instance stranded its row exactly as this finding describes, bypassing the poll clamp entirely — now gated on the same budget; and `UnitReservation::Drop` released its unit without the wake every other release path carries, which the gate turned into a 60s-per-round drain stall until fixed. Full suite 8/8 clean locally, including both new tests.

File: src/poller.rs [hunk 4]
```diff
@@ -1536,6 +1804,21 @@ impl es_entity::operation::hooks::CommitHook for ClaimHook {
...
+        // carries at most a handful of reservations from one commit, and
+        // whatever gets truncated simply releases and stays claimable by
+        // the ordinary smallest-demand-first poll.
+        let mut remaining_units = unit_budget;
+        for reservations in units_by_type.values_mut() {
+            if reservations.len() > remaining_units {
+                reservations.truncate(remaining_units);
+            }
+            remaining_units -= reservations.len();
+        }
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/poller.rs:1836]: Confirmed and fixed in e5b9852. The truncation now releases its excess reservations through the pre-existing quiet path (`UnitReservation::release`, resolves without notifying) instead of `Drop` — the budget just ran out by construction, so waking a poll that re-reads headroom blind to the hook's surviving claims was exactly backwards. The zero-budget skip keeps the waking `Drop`: there the woken poll's own budget is also zero, so it cannot over-admit — it just arms the headroom waiter (and without that wake, gate-skipped recycles stall a drainable backlog until the 60s idle fallback). New unit test pins the split: `release()` frees the unit with no wake, an unresolved drop frees it with one.

File: src/dispatcher.rs [hunk 5]
```diff
@@ -336,11 +356,62 @@ impl JobDispatcher {
...
+                        job_id = %id,
+                        attempt_no,
+                        exception.message = %e,
+                        "fail-record lost a lock conflict; retrying"
+                    );
+                    attempt_no += 1;
+                }
+                Err(e) => return Err(e),
+            }
+        }
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/dispatcher.rs]: Real — and found independently while building the finalizer consolidation: fixed in 64eb7db before this finding landed. `fail_job`'s terminal path now goes through `recycle_into_claim`, which carries the same exactly-once `recycled` guard as `BatchDispatcher::try_recycle_own_type` (a retried attempt forfeits the recycle rather than minting a second reservation for a unit the rolled-back attempt's drop already released). The code this finding targets (`fail_job_once`/`delete_execution_in_op`) no longer exists; the disposition write lives in `finalizer.rs`.

File: src/finalizer.rs [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+                // internal pool without spending an abort-retry attempt.
+                Err(e) if !use_internal => {
+                    tracing::warn!(
+                        job_ids = %Self::display_ids_of(items),
+                        exception.message = %e,
+                        "disposition write failed on the shared pool; \
+                         retrying on the internal pool"
+                    );
+                    use_internal = true;
+                }
```
[the remark this answers is not in the document]
Comment (bodymindarts) [reply src/finalizer.rs:415]: Confirmed and fixed in 0936057, both halves:

1. `finalize` now splits pre-commit from commit. The 1s deadline (and the any-error shared→internal fallback) covers ONLY acquire + writes — cancelling there is unambiguous since nothing has committed. The COMMIT itself runs uncapped and is retried only on a server-reported abort (`is_retryable_conflict` — guaranteed rollback); every other commit error is treated as ambiguous and propagates instead of retrying.

2. `finalize_in_op` now gates ALL event persistence on applied rows, the way `Complete` already was: decisions stage events on in-memory entities, every row write `RETURN`s the ids it actually transitioned, and only those ids feed the outcome, the promote registrations, and `update_all_in_op`. A rescue firing after an actually-committed attempt matches zero rows and persists nothing — closing the finding's second half (the Err-then-rescue path).
